File: mobile/lib/models/task_template.dart
class TaskTemplate {
  final int id;
  final String name;
  final String titleTemplate;
  final String? descriptionText;
  final String? icon;
  final String? color;
  final String priority;
  final String status;
  final String? category;
  final String? subcategory;
  final String? assigneeName;
  final int? dueOffsetHours;
  final int? groupId;
  final int? subgroupId;
  final List<dynamic>? checklist;
  final int useCount;

  TaskTemplate({
    required this.id,
    required this.name,
    required this.titleTemplate,
    this.descriptionText,
    this.icon,
    this.color,
    this.priority = 'medium',
    this.status = 'todo',
    this.category,
    this.subcategory,
    this.assigneeName,
    this.dueOffsetHours,
    this.groupId,
    this.subgroupId,
    this.checklist,
    this.useCount = 0,
  });

  factory TaskTemplate.fromJson(Map<String, dynamic> json) => TaskTemplate(
        id: json['id'] ?? 0,
        name: json['name'] ?? '',
        titleTemplate: json['title_template'] ?? '',
        descriptionText: json['description_text'],
        icon: json['icon'],
        color: json['color'],
        priority: json['priority'] ?? 'medium',
        status: json['status'] ?? 'todo',
        category: json['category'],
        subcategory: json['subcategory'],
        assigneeName: json['assignee_name'],
        dueOffsetHours: json['due_offset_hours'],
        groupId: json['group_id'],
        subgroupId: json['subgroup_id'],
        checklist: json['checklist'] as List?,
        useCount: json['use_count'] ?? 0,
      );
}

File: app/models/task_template.py
"""
TaskTemplate model — reusable presets for quick task creation.
"""
import json
from sqlalchemy import Column, Integer, String, Text, DateTime, Boolean
from sqlalchemy.sql import func
from app.core.database import Base


class TaskTemplate(Base):
    __tablename__ = "task_templates"

    id = Column(Integer, primary_key=True, autoincrement=True)
    name = Column(String(200), nullable=False, unique=True)
    description_text = Column(Text, nullable=True)
    icon = Column(String(10), default="📋")
    color = Column(String(7), default="#3b82f6")

    # Task field defaults
    title_template = Column(String(500), nullable=False)
    priority = Column(String(50), default="medium")
    status = Column(String(50), default="todo")
    category = Column(String(100), nullable=True)
    subcategory = Column(String(100), nullable=True)
    assignee_name = Column(String(255), nullable=True)
    label = Column(String(100), nullable=True)
    due_offset_hours = Column(Integer, nullable=True)
    group_id = Column(Integer, nullable=True)
    subgroup_id = Column(Integer, nullable=True)

    # Checklist items as JSON: [{"title":"...", "assignee_name":"..."}]
    checklist_json = Column(Text, nullable=True)

    # Meta
    is_builtin = Column(Boolean, default=False)
    is_active = Column(Boolean, default=True)
    sort_order = Column(Integer, default=0)
    use_count = Column(Integer, default=0)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    def to_dict(self):
        checklist = []
        if self.checklist_json:
            try:
                checklist = json.loads(self.checklist_json)
            except (json.JSONDecodeError, TypeError):
                pass
        return {
            "id": self.id,
            "name": self.name,
            "description_text": self.description_text,
            "icon": self.icon,
            "color": self.color,
            "title_template": self.title_template,
            "priority": self.priority,
            "status": self.status,
            "category": self.category,
            "subcategory": self.subcategory,
            "assignee_name": self.assignee_name,
            "label": self.label,
            "due_offset_hours": self.due_offset_hours,
            "group_id": self.group_id,
            "subgroup_id": self.subgroup_id,
            "checklist": checklist,
            "is_builtin": self.is_builtin,
            "is_active": self.is_active,
            "sort_order": self.sort_order,
            "use_count": self.use_count,
            "created_at": self.created_at.isoformat() if self.created_at else None,
        }
